drivers/w1/w1_gpio_delay.h

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (c) 2018 Roman Tataurov <diytronic@yandex.ru>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Driver for software driven 1-Wire using GPIO lines
+ *
+ * This driver implements an 1-Wire interface by driving two GPIO lines under
+ * software control. The GPIO pins used must be preconfigured to to a suitable
+ * state, i.e. the DQ pin as open-colector/open-drain with a pull-up resistor
+ * (possibly as an external component attached to the pin). When the DQ pin
+ * is read it must return the state of the physical hardware line, not just the
+ * last state written to it for output.
+ */
+
+#include <device.h>
+#include <errno.h>
+#include <gpio.h>
+#include <misc/printk.h>
+#include <w1.h>
+#include "w1_internal.h"
+
+#ifdef CONFIG_SOC_FAMILY_NRF
+
+#include "nrfx.h"
+
+/**
+ * Time delay for all the w1_ functions
+ *
+ * If nRF51x soc used we are using built in 16Mhz timer instead of built in
+ * ticks which are too slow for microsecond timimgs
+ *
+ * @param context [description]
+ * @param delay   delay in microseconds
+ */
+static inline void w1_delay(u32_t delay)
+{
+	/* 6us is compensation */
+	u32_t adjusted_delay = delay > 6 ? delay - 6 : 0;
+
+	nrfx_coredep_delay_us(adjusted_delay);
```

Comment:
```
warning: implicit declaration of function 'nrfx_coredep_delay_us' [-Wimplicit-function-declaration]
  nrfx_coredep_delay_us(adjusted_delay);
  ^~~~~~~~~~~~~~~~~~~~~
```

compiles but doesnt work for nrf52840_pca10056

Reply:
Tried to build with your board - got the same issue. Will check what is a trouble is.

Reply:
Try now please - should work

Reply:
It does compile now

but now Im trying to debug, I dont have a scope at hand to check it atm
```
Could not get DS18B20 device
```

Reply:
Check GPIO port name and pin specified. Also as I remember NRF ports named as GPIO_P0, but stm32 just GPIO_0. It confused a lot - in my code default GPIO naming is GPIO_0 (without P).

Reply:
Im using CONFIG_W1_GPIO_0_DQ_PIN=3, since GPIO_0 pin 0 is used for XL1

still same problem

Reply:
maybe this is relevant info:

```
warning: W1_GPIO_0 (defined at drivers/w1/Kconfig.gpio:17) was assigned the value 'y' but got the
value 'n'. You can check symbol information (including dependencies) in the 'menuconfig' interface
(see the Application Development Primer section of the manual), or in the Kconfig reference at
http://docs.zephyrproject.org/latest/reference/kconfig/CONFIG_W1_GPIO_0.html (which is updated
regularly from the master branch). See the 'Setting configuration values' section of the Board
Porting Guide as well.

warning: W1_GPIO_0_DQ_PIN (defined at drivers/w1/Kconfig.gpio:40) was assigned the value '3' but got
the value ''. You can check symbol information (including dependencies) in the 'menuconfig'
interface (see the Application Development Primer section of the manual), or in the Kconfig
reference at http://docs.zephyrproject.org/latest/reference/kconfig/CONFIG_W1_GPIO_0_DQ_PIN.html
(which is updated regularly from the master branch). See the 'Setting configuration values' section
of the Board Porting Guide as well.
```

Reply:
if i set it up with:
```
CONFIG_W1_GPIO=y
CONFIG_W1_GPIO_0=y
CONFIG_W1_GPIO_0_NAME="W1_0"
CONFIG_W1_GPIO_0_GPIO="GPIO_0"
CONFIG_W1_GPIO_0_DQ_PIN=3
```

i get not compilation errors but:
```
Unable to get 1-Wire GPIO device GPIO_1
***** Booting Zephyr OS zephyr-v1.13.0-3370-g0488f1d566 *****
Could not get DS18B20 device
```

Reply:
If I use the default demo config and override `DEFINE_W1_GPIO` in `w1_gpio.c` by:
```
DEFINE_W1_GPIO(3);
```

it still gives me:
```
***** Booting Zephyr OS zephyr-v1.13.0-3370-g0488f1d566 *****
Could not get DS18B20 device
```

Reply:
Huh - resurrected my hardware test environment :)
Got the same error now - seems zephyr API changed a lot since v1.11
Will debug it and fix

Reply:
Well here is a situation

![image](https://user-images.githubusercontent.com/36968072/55294404-889e6980-540a-11e9-9fac-1c9b766335ba.png)

On a picture we can see 2 rectangles - these are windows where bit read happened
1-st  - read 0
2-nd - read 1

So left side of window is right before read pin, but right - right after. 
Obviously MCU working too slow so it just unable to read 0 bit correctly. So finally any byte read as ``0xFF``. 

![image](https://user-images.githubusercontent.com/36968072/55294450-e0d56b80-540a-11e9-878d-8934347ba7ef.png)
 
As it was working well before on 1.11 version of Zephyr I can assume the following reasons for that

- Zephyr code itself became to work slow 
- NRF chips got some slowdown (may be working frequency reduces or something like this) - this is most probable case as I got such explanations earlier about timers.

So stuck a little for this moment :( 

Tried to increase system ticks frequency from 100 to 1000, but without success

Reply:
@diytronic first of all, thank you for all this!

What should be the procedure now? is there any specific information to raise a ticket?

Reply:
@diytronic just rebased your changes on v.1.13.0, and it behaves differently. I did that because i was trying the bme280 driver and also having problems.

https://github.com/leojrfs/zephyr/tree/rebased_v1.13

Reply:
>What should be the procedure now? is there any specific information to raise a ticket?
Not sure yet :( Played yesterday with different kernel options (ticks options, removed debugging,     turned off power save etc.). Finally get faster work, but not too much. Anyway my code still does not work.

Checked NRF specification - MCU always work on 16 MHz so seems problem is not in specific chip settings. 

> Rebased your changes on v.1.13.0, and it behaves differently
Another one thing I discovered is GPIO work differently. So think may be issue is here. Used the following GPIO settings 
```
gpio_pin_configure(context->gpio, context->dq_pin, (GPIO_PUD_PULL_UP | GPIO_DIR_OUT));
```
It was work perfectly on 1.11, but now I got some issues.

- right after pin configuration pins state is low (very strange for pull up pin) so fixed it to additional call to put it to 1
- it always read 1 - seems pull up always read as 1

So tried temporary switch to ``GPIO_DIR_IN`` before read and it work, but currently only one time. Next time always read 1 (may be it is timing issue described above - not sure yet). 

So this is definitely timings or some issues with gpio. Symptom is - GPIO always read 1.

Reply:
BTW found this - related

https://lists.zephyrproject.org/g/devel/message/5654?p=,,,20,0,0,0::relevance,,GPIO,20,2,100,29656220

Reply:
Hmm - bad news

> Yes there is no bitbanding on the nRF series chips. They are only mapped as bits in the 32 bit PORT registers. There is no way of getting one single address to read/write from for a single pin. 

https://devzone.nordicsemi.com/f/nordic-q-a/31291/nrf52-address-of-single-pin-s-input-output-register/123609#123609

>  the possibility to use the 51's GPIOs as both input and output simultaneously was actually an unintentional "feature" and that it is no longer possible on the 52.

https://devzone.nordicsemi.com/f/nordic-q-a/9508/problem-of-gpio-configuration-difference-between-nrf51822-and-nrf52832/35112#35112

So finally seems bitbang 1-wire on nrf chips isnt possible :(

Reply:
oh, good catch... 
Most of your work is still useful to implement a `DS2482` driver I guess.

Thanks